### ISSUE TITLE
Infallible functions

### ISSUE THREAD

**Issue by encukou** [capi-workgroup/problems#20]
Several functions currently never fail, and don't have a way to report failure. This blocks development of features that need an error condition.

**Comment by markshannon** [comment]
That doesn't seem like it is infallible, that seems like the API is flawed.
Not having a means for reporting failure is not the same as infallibility.

There are infallible functions, e.g. `PyTuple_Size()`

**Comment by encukou** [comment]
Yes, the API is flawed. Same with `PyStructSequence_SetItem`, for example.

**Comment by gvanrossum** [comment]
Wouldn't `PyTuple_Size` fail if you pass it a non-tuple?

In a sense even (especially?!) `DECREF` has this flaw. There's nowhere to report errors incurred during finalization.

**Comment by markshannon** [comment]
> Wouldn't `PyTuple_Size` fail if you pass it a non-tuple?

You're right, it does. My mistake.
`PyTuple_Size` takes `PyObject *` not `PyTupleObject *` so it needs to do a type check.
The inline version `PyTuple_GET_SIZE` doesn't fail.

**Comment by encukou** [comment]
IMO, so few API functions are infallible (in all possible implementations) that any guidelines should say there should be a way to report errors. Any exceptions need to be carefully thought out.
Like in #21.

When failure is impossible in the current implementation and the check is too expensive, inline the function and let the compiler elide the check.

**Comment by encukou** [comment]
To 3.13, yes.
If it's void now, changing the return value to int isn't an ABI break on common/supported architectures, so in practice it should be safe to do in 3.12 too. But maybe not worth convincing the RM :)
For older releases, I'd be more careful, and only change the docs mention having to call `PyErr_Occurred()`.

But for a function that previously didn't set an exception at all, adding such an error case would, IMO, be an API break: there should be a new function and the old one should be deprecated.

**Comment by iritkatriel** [comment]
Labelling this as fixable, let's work on it in https://github.com/python/cpython/issues/105184.

**Comment by erlend-aasland** [comment]
We can work on guidelines for new APIs here: https://github.com/python/devguide/issues/1122

**Comment by vstinner** [comment]
``PyFrame_FastToLocals()`` has to allocate a dict but its return type is ``void``: it cannot report errors. So I added ``PyFrame_FastToLocalsWithError()`` which returns -1 on error (if an exception was set).

``PyDict_GetItem()`` has a similar story: ``PyDict_GetItemWithError()`` was added.

It's convenient to have functions which cannot fail: it avoids to have to write boring (complex to write, easy to mess up error handling code, etc.) error code path. For example, getter functions like ``PyFrame_GetCode()`` "just" reads an attribute of a structure, so it cannot fail, right?

Well, then things become more complicated :-) Until Python 3.10, ``PyFrame_GetBack()`` just read a structure member. But in Python 3.11, if the frame has no concrete Python frame object, it has to allocate memory and the allocation can fail. On memory allocation failure, PyFrame_GetBack() returns NULL which means "no previous frame" :-(

**Comment by encukou** [comment]
Also, emitting a warning can cause errors (`-Werror`, allocation, etc.). So, if an API might ever need to raise a runtime `DeprecationWarning` (e.g. for a functionality change in *some* cases, or before removal from a stable ABI), it can't be `void`.

**Comment by vstinner** [comment]
When a function evolved and then can fail but the API cannot report error, a possible workaround is to *log* the error with ``sys.unraisablehook``.

**Comment by encukou** [comment]
Proposed guideline issues:
- https://github.com/capi-workgroup/api-evolution/issues/5
- https://github.com/capi-workgroup/api-evolution/issues/13
